src/types/Writable.ts

```diff
@@ -32,5 +38,20 @@ export type Writable<T> = T extends WritablePrimitive
 			: T extends Set<infer X>
 				? Set<Writable<X>>
 				: T extends Object
-					? { [K in WritableKeys<T>]: Writable<T[K]> }
+					? {
+							// Capture required keys
+							[K in keyof T & RequiredKeys<T> & WritableKeys<T>]: Writable<
+								T[K]
+							>;
+						} & {
+							// Capture optional keys
+							[K in keyof T & OptionalKeys<T> & WritableKeys<T>]?: Writable<
+								T[K]
+							>;
+						} extends infer O
+						? {
+								// Flatten the two types (required and optional) into a single combined type
+								[K in keyof O]: O[K];
+							}
+						: never
```

Comment:
Label the [intersection type](https://www.typescriptlang.org/docs/handbook/2/objects.html#intersection-types) as `O`, so that we can do an identity pass on it.

Is this necessary? I was able to delete this clause and it still compiles / builds / passes ESLint:

```suggestion
						}
```

Reply:
Aha, I see that now too. Thank you for walking through this with me!